internal/upstreamldap/upstreamldap.go

```diff
@@ -836,3 +860,68 @@ func getDomainFromDistinguishedName(distinguishedName string) (string, error) {
 	}
 	return strings.Join(domainComponents[1:], "."), nil
 }
+
+func PwdUnchangedSinceLogin(entry *ldap.Entry, attributes provider.StoredRefreshAttributes) error {
+	authTime := attributes.AuthTime
+	pwdLastSetWin32Format := entry.GetAttributeValues(PwdLastSetAttribute)
+	if len(pwdLastSetWin32Format) != 1 {
+		return fmt.Errorf("expected to find 1 value for %s attribute, but found %d", PwdLastSetAttribute, len(pwdLastSetWin32Format))
+	}
+	// convert to a time.Time
+	pwdLastSetParsed, err := win32timestampToTime(pwdLastSetWin32Format[0])
+	if err != nil {
+		return err
+	}
+
+	// if pwdLastSet > authTime, that means that the password has been changed since the initial login.
+	// return an error so the user is prompted to log in again.
+	if pwdLastSetParsed.After(authTime) {
```

Comment:
authtime in this case is the time the user logged in (e.g. "this morning"). PwdLastSet is whenever the AD password was last changed, which could be weeks or months ago. The only thing that should trigger a failure would be if you logged in in the morning and changed your password in the afternoon.

Reply:
Ah I understand.

This construction is not how I would expect this check to be implemented.  I would expect us to record the `pwdLastSet` on login and assert that it is unchanged on refresh.  I think that would allow us to drop the date parsing code altogether.  More importantly, it reflects how such a check could be implemented for generic LDAP (i.e. give me list of attributes and I will fail refresh when they change from what I initially recorded).

Reply:
That makes sense. I originally implemented it this way because it is one less thing to store, since we already keep track of the time of authentication. But it does mean the date parsing is more complicated 😬. Should I change it?

Reply:
Since we are not in a rush, I would go ahead and change it since it sets us up for having a better generic LDAP refresh check in the future.  And I think we could drop the date parsing code which would be nice.